File: src/main/java/model/pieces/DirectionPiece.java
package model.pieces;

import model.util.BoardMap;
import model.util.Colour;
import model.util.Offset;
import model.util.Position;

import java.util.Collection;
import java.util.LinkedList;

/**
 * A piece that attacks in a line a (or several) directions (ex. Queen, Rook, Bishop)
 */
abstract class DirectionPiece extends Piece {
    DirectionPiece(Colour colour) {
        super(colour);
    }

    @Override
    Collection<Position> generatePossibleDestinations(BoardMap board, Position start) {
        Collection<Position> positions = new LinkedList<>();

        //For each direction
        for (Offset direction : getDirections()) {
            Position end = start.shift(direction);

            //Loop through all the positions in that line/direction
            while (end.isValid()) {
                Piece piece = board.getPiece(end);

                if (piece == null) positions.add(end); //If the square (position) is empty -> can move
                else {
                    //If piece is other colour can eat
                    if (piece.getColour() != colour) positions.add(end);

                    //Cannot go past a piece since it is blocking the line
                    break;
                }

                end = end.shift(direction);
            }
        }

        return positions;
    }

    /**
     * @return the list of directions the piece can attack
     */
    abstract Offset[] getDirections();
}

File: engine/src/main/java/engine/pieces/DirectionPiece.kt
package engine.pieces

import engine.GameData
import engine.util.Colour
import engine.util.Position
import java.util.*

/**
 * A pieceMap that attacks in a line a (or several) directions (ex. Queen, Rook, Bishop)
 */
internal abstract class DirectionPiece(colour: Colour) : Piece(colour) {

    /**
     * the list of directions the pieceMap can attack
     */
    internal abstract val directions: Array<Position>

    override fun generatePossibleDestinations(gameData: GameData, start: Position): Collection<Position> {
        //WARNING if parameter start is removed pawn promoted to queen might not work since queen will not recognize itself on the pieceMap
        val positions = LinkedList<Position>()

        //For each direction
        for (direction in directions) {
            var end = start + direction

            //Loop through all the positions in that line/direction
            while (end.isValid) {
                val piece = gameData.pieceMap[end]

                if (piece == null) positions.add(end) //If the square (position) is empty -> can move
                else {
                    //If pieceMap is other colour can eat
                    if (piece.colour != this.colour) positions.add(end)

                    //Cannot go past a pieceMap since it is blocking the line
                    break
                }

                end += direction
            }
        }

        return positions
    }
}
